module/platform/NUsense/HardwareIO/src/NUSenseParser.cpp

```diff
@@ -0,0 +1,61 @@
+#include "NUSenseParser.hpp"
+
+namespace module::platform::NUsense {
+
+    /**
+     * Read a little endian value from a byte array
+     *
+     * @tparam T The type of the value to read
+     *
+     * @param ptr The pointer to the first byte of the value
+     *
+     * @return The value read from the byte array
+     */
+    template <typename T>
+    T read_le(const uint8_t* ptr) {
```

Comment:
This needs to include a max array length.

Also, just use htons/ntohs (or whatever variant is appropriate) and implement a [64-bit version](https://stackoverflow.com/questions/3022552/is-there-any-standard-htonl-like-function-for-64-bits-integers-in-c)

Reply:
Including max array length for this is the responsibility of the caller which is why it's a private cpp only function not a public one for anyone to use. All the usages of it in the functions below have had appropriate length checks on the buffer to ensure no out of bounds memory access happens.

htons and ntohs only work for dealing with big endian values, not little endian values since they interact with "native" and "network" byte order only. So reading little-endian values form the network and converting them to native doesn't work with them.

That said, after putting this code through godbolt, it seems that the compiler isn't as smart as I thought when it comes to converting to appropriate instructions.

I made a bunch of implementations and it seems to only recognise what is going on here in a few cases.
I think either the bitwise version of the recursive template would be best given that it recognises those two correctly. 
https://godbolt.org/z/1Wof68xGq

Reply:
Just dump this function and use a reinterpret cast, the likelihood of ever running on a big endian system is so slim that it doesn't bare accounting for